ENIGMAsystem/SHELL/Universal_System/Extensions/ttf/ttf.cpp

```diff
@@ -37,32 +45,122 @@ class FontManager {
   static FT_Library library;
 };
 
+struct GlyphPair {
+  GlyphPair(unsigned index = 0, unsigned area = 0) : index(index), area(area) {}
+  unsigned index;
+  unsigned area;
+};
+
 FT_Library FontManager::library;
 static bool FreeTypeAlive = FontManager::Init();
 
+std::vector<std::string> fontSearchPaths;
+
+const std::map<std::string, std::string> fontFallbacks = {
+  {"Arial", "LiberationSans"},
+  {"Times New Roman", "LiberationSerif"},
+  {"Courier New", "LiberationMono"} 
+};
+
 }
 
+
 namespace enigma_user {
   
   using enigma::rect_packer::pvrect;
   using enigma::rect_packer::rectpnode;
+  using enigma::GlyphPair;
+  using enigma::fontSearchPaths;
+  using enigma::fontFallbacks;
+  
+  void font_add_search_path(std::string path) {
+    fontSearchPaths.push_back(path);
+  }
   
-  struct GlyphPair {
-    GlyphPair(unsigned index = 0, unsigned area = 0) : index(index), area(area) {}
-    unsigned index;
-    unsigned area;
-  };
+  std::string font_find(std::string name, bool bold, bool italic, bool fallback) {
+    
+    // Unix Style naming Font-[Bold][Italic][BoldItalic][Regular].ttf
+    // Windows Style namings font[b or bd or B][i or it or I][bi or z or S].ttf (often all lowercase)
+    std::vector<std::string> possibleNames;
+    std::string suffix;
+    
+    // Unix Best Match
+    if (bold) suffix = "Bold";
+    if (italic) suffix += "Italic";
+    if (suffix.empty()) suffix = "Regular";
+    possibleNames.push_back(name + "-" + suffix + ".ttf");
+    
+    auto addWinNames = [&](std::string s1, std::string s2, std::string s3) {
+      std::string n = name + "-" + s1 + ".ttf";
+      possibleNames.push_back(n);
+      std::transform(n.begin(), n.end(), n.begin(), ::tolower);
+      possibleNames.push_back(n);
+      
+      n = name + "-" + s2 + ".ttf";
+      possibleNames.push_back(n);
+      std::transform(n.begin(), n.end(), n.begin(), ::tolower);
+      possibleNames.push_back(n);
+      
+      n = name + "-" + s3 + ".ttf";
+      possibleNames.push_back(n);
+    };
+    
+    // Win32 Best Match
```

Comment:
The file names on win32 font match the names of the font for whatever reason so we need some kind of map for that too